pdbtools/pdb_selaltloc.py

```diff
@@ -106,66 +108,111 @@ def check_input(args):
-
-    # We have to iterate multiple times
-    atom_prop = {}  # {atom_uid: (lineno, prop)}
+    atom_prop = collections.defaultdict(list)
     atom_data = []
+    anisou_lines = {}  # map atom_uid to lineno
+    ignored = set()
 
     # Iterate over file and store atom_uid
-    records = ('ATOM', 'HETATM')
-    for line in fhandle:
+    records = ('ATOM', 'HETATM', 'ANISOU')
+    for lineno, line in enumerate(fhandle):
 
         atom_data.append(line)
 
         if line.startswith(records):
-            atom_uid = (line[12:16], line[17:26])
-            atom_full_uid = line[12:26]
-            prop = get_prop(line)
-
-            if atom_uid in atom_prop:
-                atom_prop[atom_uid].append((atom_full_uid, prop))
+            # Sometimes altlocs are used between different residue names.
+            # See 3u7t (residue 22 of chain A). So we ignore the resname below.
+            atom_uid = (line[12:16], line[20:26])
+
+            # ANISOU records do not have occupancy values.
+            # To keep things simple, we map ANISOU to ATOM/HETATM records
+            if line.startswith('ANISOU'):
+                anisou_lines[lineno - 1] = lineno
+                ignored.add(lineno)  # we will fix this below
             else:
-                atom_prop[atom_uid] = [(atom_full_uid, prop)]
+                occ = float(line[54:60])
+                atom_prop[atom_uid].append((lineno, occ))
+
+    # Iterate and pick highest occupancy for each atom.
+    for atom_uid, prop_list in atom_prop.items():
+        prop_list.sort(key=operator.itemgetter(1), reverse=True)
+
+        lineno = prop_list[0][0]
+
+        # Edit altloc field(s)
+        line = atom_data[lineno]
+        atom_data[lineno] = line[:16] + ' ' + line[17:]
 
-    # Filter atom_prop
-    sel_atoms = set()
-    for key, prop_list in atom_prop.items():
-        selected = sel_prop(prop_list)
-        sel_atoms.add(selected[0])  # atom_full_uid
+        if lineno in anisou_lines:
+            anisou_lineno = anisou_lines[lineno]
+            line = atom_data[anisou_lineno]
+            atom_data[anisou_lineno] = line[:16] + ' ' + line[17:]
+            ignored.discard(anisou_lineno)
 
-    # Iterate again and yield the right one
-    records = ('ATOM', 'HETATM', 'ANISOU')  # we can filter ANISOU too
+        ignored.update(p[0] for p in prop_list[1:])
+
+    # Now yield
     for lineno, line in enumerate(atom_data):
+        if lineno in ignored:
+            continue
+
+        yield line
+
+
+def select_by_altloc(fhandle, selloc):
+    """Picks one altloc when atoms have more than one.
+
+    If the specified altloc (selloc) is not present for this particular atom,
+    outputs all altlocs. For instance, if atom X has altlocs A and B but the
+    user picked C, we return A and B anyway. If atom Y has altlocs A, B, and C,
+    then we only return C.
+    """
+
+    # We have to iterate multiple times
+    atom_prop = collections.defaultdict(list)
```

Comment:
Despite using `collections` is more verbose and pythonic, because pdb-tools will very likely use in pipelines, and to maintain one of the main ideas of the project, use built-ins before stdlib, I suggest reverting to dict and `.setdefault` since the simpler approach is 3x faster than `defaultdict`.

```python
%%timeit
a = collections.defaultdict(list)
a['b'].append(1)
```
```
383 ns ± 10.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```


```python
%%timeit
b = {}
c = b.setdefault('b', [])
c.append(1)
```

```
127 ns ± 2.51 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```